apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfSubscriber.cs

```diff
@@ -0,0 +1,723 @@
+                _flowControlSettings.MaxOutstandingElementCount ?? long.MaxValue, RegisterTask, _taskHelper);
+            // Start all subscribers
+            var subscriberTasks = _clients.Select(client =>
+            {
+                var singleChannel = new SingleChannel(_subscriptionName, client, handlerAsync,
+                    _autoExtendInterval, _modifyDeadlineSeconds, _maxAckExtendCount, flow,
+                   RegisterTask, _scheduler, _globalSoftStopCts, _globalHardStopCts, _taskHelper);
+                return singleChannel.StartAsync();
+            }).ToArray();
+            // Setup finsh task
+            _taskHelper.Run(async () =>
+            {
+                var task = await _taskHelper.ConfigureAwait(_taskHelper.WhenAny(subscriberTasks));
+                if (task.IsFaulted)
+                {
+                    _globalHardStopCts.Cancel();
+                }
+                // Wait for all subscribers to stop
+                var exception = await _taskHelper.ConfigureAwaitHideErrors(_taskHelper.WhenAll(subscriberTasks));
+                // Wait for all registered Tasks to stop
+                await _taskHelper.ConfigureAwaitHideErrors(_taskHelper.WhenAll(registeredTasks.Locked(() => registeredTasks.ToArray())));
+                // Call shutdown function
+                if (_shutdown != null)
+                {
+                    await _taskHelper.ConfigureAwaitHideErrors(_shutdown);
+                }
+                // Return final result
+                var exceptions = ((exception as AggregateException)?.Flatten()?.InnerExceptions) ?? Enumerable.Repeat(exception, exception == null ? 0 : 1);
+                if (exceptions.Any())
+                {
+                    _mainTcs.SetException(exceptions);
+                }
+                else if (_globalHardStopCts.IsCancellationRequested)
+                {
+                    _mainTcs.SetCanceled();
+                }
+                else
+                {
+                    _mainTcs.SetResult(0);
+                }
+            });
+            return _mainTcs.Task;
+        }
+
+        /// <inheritdoc />
+        public override Task StopAsync(CancellationToken hardStopToken)
+        {
+            lock (_lock)
+            {
+                GaxPreconditions.CheckState(_mainTcs != null, "Can only stop a started instance.");
+                GaxPreconditions.CheckState(!_globalSoftStopCts.IsCancellationRequested, "Can only stop an instance once.");
+                _globalSoftStopCts.Cancel();
+            }
+            var registration = hardStopToken.Register(() => _globalHardStopCts.Cancel());
+            // Do not register this Task, it completes *after* mainTask
+            _taskHelper.Run(async () => await _taskHelper.ConfigureAwaitWithFinally(_mainTcs.Task, () => registration.Dispose()));
+            return _mainTcs.Task;
+        }
+
+        private class Flow
```

Comment:
A doc comment here would be very useful. I don't really know what this means yet...